pkg/kv/kvserver/kvflowcontrol/replica_rac2/admission.go

```diff
@@ -0,0 +1,277 @@
+//
+// For C1, consider the following sequence:
+//   - e10 NormalPri not stepped successfully
+//   - e10 LowPri override stepped successfully
+//   - e10 emerges in handleRaftReady: we will use NormalPri, though should have
+//     used LowPri.
+//
+// TODO(sumeer): consider increasing the accuracy by returning information
+// from step.
+type lowPriOverrideState struct {
+	// intervals for which information has been provided.
+	//
+	// A call to getEffectivePriority for index i causes a prefix of indices <=
+	// i to be discarded.
+	intervals []interval
+	// Highest term observed so far.
+	leaderTerm uint64
+}
+
+// Represents [first, last].
+type interval struct {
+	first          uint64
+	last           uint64
+	lowPriOverride bool
+}
+
+// sideChannelForLowPriOverride is called on follower replicas, and specifies
+// information for [first, last]. An index can never be queried using
+// getEffectivePriority without it being first specified in
+// sideChannelForLowPriOverride.
+//
+// A new leader can overwrite previous log entries. And the existing leader
+// can retransmit entries, with a LowPri override. In the case of an existing
+// leader we only extend the existing state, as discussed in the correctness
+// comment above.
+//
+// INVARIANT: first <= last.
+func (p *lowPriOverrideState) sideChannelForLowPriOverride(
+	leaderTerm uint64, first, last uint64, lowPriOverride bool,
+) {
+	newLeader := false
+	if leaderTerm < p.leaderTerm {
+		return
+	} else if leaderTerm > p.leaderTerm {
+		newLeader = true
+		p.leaderTerm = leaderTerm
+	}
+	n := len(p.intervals)
+
+	// Given either n == 0, or n-1 is the last element and p.intervals[n-1].last
+	// < first.
+	extendOrAppend := func() {
+		if n != 0 && p.intervals[n-1].lowPriOverride == lowPriOverride &&
+			p.intervals[n-1].last+1 >= first {
+			// Extend it.
+			p.intervals[n-1].last = last
+			return
+		}
+		p.intervals = append(p.intervals,
+			interval{first: first, last: last, lowPriOverride: lowPriOverride})
+	}
+	if newLeader {
+		// Drop all intervals starting at or after the first index. Do it from the
+		// back, so that the append-only case is the fast path. When a suffix of
+		// entries is overwritten, the cost of this loop is an amortized O(1).
+		for ; n != 0 && p.intervals[n-1].first >= first; n-- {
+		}
+		p.intervals = p.intervals[:n]
+		// INVARIANT: keep > 0 => p.intervals[keep-1].first < first.
+
+		// Adjust the last interval if it overlaps with the one being added.
+		if n != 0 && p.intervals[n-1].last >= first {
+			// Truncate the last interval.
+			p.intervals[n-1].last = first - 1
+		}
+		extendOrAppend()
+		return
+	}
+	// Common case: existing leader.
+	if n > 0 {
+		if p.intervals[n-1].last >= last {
+			return
+		}
+		// INVARIANT: p.intervals[n-1].last < last, so p.intervals[n-1].last + 1 <= last.
+		if p.intervals[n-1].last >= first {
+			first = p.intervals[n-1].last + 1
+		}
+	}
+	extendOrAppend()
+}
+
+func (p *lowPriOverrideState) getEffectivePriority(
```

Comment:
Can we decouple the data structure from the priority override logic at the bottom of this func? For what I see, the data structure can be a list of overridden intervals + lastIndex, and only needs to return true of false (from some `shouldOverride` method). Then there would be a standalone `getEffectivePriority` function which applies the true/false override to the original entry priority.

Decoupling the two makes things clear and more easily testable (no combinatorics).

Also, arguably, this struct does not need to know that the override is to a low priority - it's just concerned with a collection of intervals.